### ISSUE TITLE
Single method causing "incompatible signatures" error

### ISSUE THREAD

**Issue by brandondiamond** [dart-lang/sdk#42248]
Environment:
 - Dart 2.8.4
 - built_value 7.1.0
 - Mac OS X 10.15.4

Summary:

I have a generic method with two recursive type arguments. The first is a built value and the second is the corresponding builder.

As a further complication, this method lives in a generic mixin with its own similar (but distinct) type arguments as well as a superclass constraint:

```
mixin ExampleMixin<
    T extends built_value.Built<T, S>,
    S extends built_value.Builder<T, S>> on BaseClass {

 void broken<V extends built_value.Built<V, W>,
          W extends built_value.Builder<V, W>>({V initial,
      W initialBuilder}) {}

}
```

This mixin is later used like so:

```
class ExampleClass extends BaseClass with ExampleMixin<Value, ValueBuilder> {
  void someMethod() {
    broken<AnotherValue, AnotherValueBuilder>();
  }
}
```

This breaks compilation with the following somewhat surprising error:

```
Error: Class 'BaseClass with ExampleMixin<Value, ValueBuilder>' inherits multiple members named 'broken' with incompatible signatures.
Try adding a declaration of 'broken' to 'BaseClass with ExampleMixin<Value, ValueBuilder>'.
class ExampleClass extends BaseClass
      ^
[... Code path ...]: Context: This is one of the inherited members.
  void broken<V extends built_value.Built<V, W>,
       ^^^^^^^
[... Identical path ...]: Context: This is one of the inherited members.
  void broken<V extends built_value.Built<V, W>,
       ^^^^^^^
```

Replacing the method's type signature with a less precise alternative fixes the bug:

```
mixin ExampleMixin<
    T extends built_value.Built<T, S>,
    S extends built_value.Builder<T, S>> on BaseClass {

 void broken<V extends built_value.Built<dynamic, dynamic>,
          W extends built_value.Builder<dynamic, dynamic>>({V initial,
      W initialBuilder}) {}

}
```

**Comment by devoncarew** [comment]
@lrhn @eernstg - is this expected for mixins or an issue w/ the CFE?

**Comment by eernstg** [comment]
I'm assuming that there is no declaration of `broken` in `BaseClass`? If so then this is not expected, there is no notion of incompatible member signatures that would kick in when only one superinterface has that member.

**Comment by lrhn** [comment]
I agree that if there is no `broken` member in `BaseClass`, then there should be no error here.

It might be a problem arising from inheriting the member implementation from the mixin application as well as from the mixin interface, and the compiler failing to see that the two recursive type bounds are the same type.

In any case, the warning is definitely not useful - you cannot add a method to a mixin application.
It might be an issue that `class C extends B with M implements I { ...}` creates a class `B with M` which needs to be valid, without there being any way to insert a conflict resolving declaration. Users might think that members declared in `C` would resolve the conflict.

**Comment by eernstg** [comment]
I don't think we can have a situation with `B with M` as such which justifies a new treatment of conflicts (and perhaps a new treatment of where conflict resolving declarations can be located): `B with M` could have an incorrect override error, but it wouldn't have a symmetric conflict involving `B` and `M` because `B` is the superclass of `B with M`. But if there is an incorrect override error then `M` should not be applied to `B`.

**Comment by lrhn** [comment]
Ah, yes. `B with M` has `B` as superclass and `M` as superinterface, but it also has the implementation from `M` as a direct member overriding both. If there is any error, it's an invalid override, not a conflict (well, except if the `B` and `M` interfaces have member differences that are not declared in `M`.

**Comment by brandondiamond** [comment]
I did some more experimentation and found the trigger -- something I failed to mention in either of my earlier posts.

The broken method accepts a function as a parameter. This function expects two arguments: the first corresponding to the mixin's type parameter and the second corresponding to the method's type parameter. Something about "crossing the streams" is causing things to fall apart.

**From this, I was able to put together a reproduction:**
https://gist.github.com/brandondiamond/423c87736f4d2fe506afcbb13550f24e

This produces the described error when run in dartpad. Interestingly, the analyzer has no issues with the code.

(Why am I doing this? The purpose of this method is to merge two built values: the one used with the mixin, and another that results from calling this problematic method. The callback at the root of this bug does the work of merging the two built values and so deals with generics bound in the method and in the containing mixin).

**Comment by lrhn** [comment]
Excellent reproduction. 
I see no problem with the code (and neither does the analyzer), so let's see if this is a front-end error.
Definitely a corner case.

**Comment by brandondiamond** [comment]
Thank you so much for digging into this -- I've very grateful.

I was able to further simplify the reproduction, though this is a bit more obtuse than my first example:

https://gist.github.com/brandondiamond/a5359294a67ea186ecb7773f8764feea
